docs/architecture/key-management-services.md

```diff
@@ -108,4 +112,143 @@ There can be some variations in the support of these crypto primitives, hashing
+
+Given the [**principles**](#kms-principles) we have defined, particularly the `isolation` principle, we need to consider a way to achieve isolation in a non-hardware environment. 
+
+Because we are dependent on the run-time environment to provide isolation, we should consider OS'es specifics and their capabilities in order to achieve a secure design.
+
+#### Standard Web3 Design
+
+```mermaid
+    C4Context
+        title "Non isolated KMS & App"
+        Boundary(b0, "Android", "OS") {
+            Boundary(b1, "Wallet", "Main process") {
+                System(app, "OWF Engine", "wallet")
+                System(kms, "KMS", "Key Management Service")
+
+            }
+
+            BiRel(app, kms, "uses", "Shared Memory")
+        }
+
+        UpdateRelStyle(app, kms, $textColor="green", $lineColor="blue", $offsetX="10")
+        UpdateLayoutConfig($c4ShapeInRow="1", $c4BoundaryInRow="0")
+```
+
+##### Problem with Shared Run-time space
+- Shared resources
+- Permissions to the app open attack surface
+- No compartmentalization of a breach or zero-day exploit
+Non compliant with some security industry standards and regulations where separation of duties is a must
+- Breaks isolation principle
+
+
+##### Precedence for software based enclaves
+
+There's already precendence outside of web3 for software-only based enclaves. For example:
+- [HashiCorp Vault](https://www.vaultproject.io/)
+- [softHSM](https://www.opendnssec.org/softhsm/)
+- [Apple CryptoKit](https://developer.apple.com/documentation/cryptokit)
+    - note: Only NIST P-256 is hardware backed. Curve25519 and EdDSA are **NOT** hardware backed.
+
+
+#### Isolation Strategies
+
+- Virtualization
+- Containers
+- Separate Applications
+- Processes & Services
+- Trusted Execution Environments
+
+##### High Level Conceptual Design
+
+```mermaid
+    C4Context
+        title "Isolation KMS & App"
+        Boundary(b0, "Android", "OS") {
+            Boundary(b1, "Wallet", "Main process") {
+                System(app, "OWF Engine", "wallet")
+            }
+
+            Boundary(b2, "KMS", "Isolated Service") {
+                System(kms, "KMS", "Key Management Service")
+            }
+
+            BiRel(app, kms, "uses", "IPC")
+        }
+
+        UpdateRelStyle(app, kms, $textColor="green", $lineColor="blue", $offsetX="0")
+        UpdateLayoutConfig($c4ShapeInRow="1", $c4BoundaryInRow="0")
+
+```
+
+#### Key Material Handling
+
+- Use device bound hardware for encryption / decryption
+- At rest encryption with native support through TEE & phone’s native HSM 
+- At run-time decrypt only necessary keys
+- The application is blind to any private key material and should only request operations from the KMS. Operations can be (not excluded to):
+    - Key-gen
+    - Get-publicKey
+    - Sign
+    - Key-derive
+    - Key-import 
+    - etc
```

Comment:
I am sure we are not the first to propose such a separation.
Can we have examples of other systems working this way?
Can we learn from their experience?
Can we study and reuse their APIs?